ssg/build_remediations.py

```diff
@@ -218,26 +220,94 @@ def parse_from_file_without_jinja(file_path):
-    """
-    Process a fix, adding it to fixes iff the file is of a valid extension
-    for the remediation type and the fix is valid for the current product.
+class Remediation(object):
+    def __init__(self, env_yaml, resolved_rules, product, file_path, fix_name, remediation_type):
+        self.env_yaml = env_yaml
+        self.resolved_rules = resolved_rules
+        self.product = product
+        self.file_path = file_path
+        self.fix_name = fix_name
+        self.remediation_type = remediation_type
 
-    Note that platform is a required field in the contents of the fix.
-    """
+    def process(self, fixes):
+        """
+        Process a fix, adding it to fixes iff the file is of a valid extension
+        for the remediation type and the fix is valid for the current product.
 
-    if not is_supported_filename(remediation_type, file_path):
-        return
+        Note that platform is a required field in the contents of the fix.
+        """
+
+        if not is_supported_filename(self.remediation_type, self.file_path):
+            return
+
+        result = parse_from_file_with_jinja(self.file_path, self.env_yaml)
+
+        if not result.config['platform']:
+            raise RuntimeError(
+                "The '%s' remediation script does not contain the "
+                "platform identifier!" % (self.file_path))
+
+        if is_applicable_for_product(result.config['platform'], self.product):
+            fixes[self.fix_name] = result
+
+        return result
+
+
+class BashRemediation(Remediation):
+    def __init__(self, env_yaml, resolved_rules, product, file_path, fix_name):
+        super(BashRemediation, self).__init__(
+            env_yaml, resolved_rules, product, file_path, fix_name, "bash")
 
-    result = parse_from_file_with_jinja(file_path, env_yaml)
 
-    if not result.config['platform']:
-        raise RuntimeError(
-            "The '%s' remediation script does not contain the "
-            "platform identifier!" % (file_path))
+class AnsibleRemediation(Remediation):
+    def __init__(self, env_yaml, resolved_rules, product, file_path, fix_name):
+        super(AnsibleRemediation, self).__init__(
+            env_yaml, resolved_rules, product, file_path, fix_name, "ansible")
 
-    if is_applicable_for_product(result.config['platform'], product):
-        fixes[fix_name] = result
+    def process(self, fixes):
+        result = super(AnsibleRemediation, self).process(fixes)
+
+        rule_path = os.path.join(
+            self.resolved_rules, self.fix_name + ".yml")
+        if not os.path.isfile(rule_path):
+            msg = ("Ansible snippet for rule {rule_id} "
+                   "doesn't have respective rule YML at {rule_fname}"
+                   .format(rule_id=os.path.splitext(self.fix_name)[0], rule_fname=rule_path))
+            print(msg, file=sys.stderr)
+            return result
+
+        import ssg.ansible
+        from ssg import build_yaml
```

Comment:
Can the imports be at the beginning of file?

Reply:
Unfortunately not, there is an import loop, and late imports are the only way how to make it work. The proper solution would be to unify the two remediation classes and straightening the dependency graph as a consequence.